### ISSUE TITLE
t helper does not accept a hash object directly

### ISSUE THREAD

**Issue by SemaLuna** [ember-intl/ember-intl#1566]
- [ ] I am on the [latest](https://github.com/ember-intl/ember-intl/releases/latest) ember-intl version
  - Not on the latest, but I believe this still applies to the latest version (looking at the code)
- [x] I have searched the [issues](https://github.com/ember-intl/ember-intl/issues) of this repo and believe that this is not a duplicate

### Environment

- Ember Version: 3.20.5
- Ember CLI Version: 3.15.2
- Ember Intl Version: 5.5.0
- Browser(s): Any
- Node Version: 12.18.4

### Steps to Reproduce

If I have a translation with multiple placeholders, 

`"myTranslation": "You can either {option1} or {option2}"`

I need to pass them individually, like so:

`{{t myTranslationPath option1='run' option2='hide'}}`

It would be ideal if we could pass the options hash directly, like so:

`{{t myTranslationPath optionsHash}}`

where the `optionsHash` is 

`{option1="run", option2="hide"}`

However - because of how helpers work, any non-named arguments are passed as an array, and those are not picked up by the 't' helper (just get an error saying that I did not pass the placeholders).

In the above example, it's just two arguments - but imagine the case where we have variable translation strings, and thus variable arguments - it's extra legwork that should not be necessary.

Not sure if this is a bug or considered an enhancement (or if there's a good workaround that I am missing).

**Comment by nolaneo** [comment]
I have also run into this in the past. My workaround was to just make a `t2` helper which extends the original but allows for a context hash.

Note this is for version 4.2.0 but the approach would still work in 5.X.X

```js
// This helper simply allows us to spread the inputs
// {{t2 'translation-key' contextObject otherOptionKey='yes'}}

import T from 'my-app-name/helpers/t';
import { isEmpty } from '@ember/utils';

export default T.extend({
  compute([value, context], options) {
    if (isEmpty(value)) {
      if (options.allowEmpty ?? this.allowEmpty) {
        return;
      }

      if (typeof value === 'undefined') {
        throw new Error(`${this} helper requires value attribute.`);
      }
    }

    return this.format(value, options, context);
  },

  format(key, options, context = {}) {
    return this._super(key, { ...context, ...options });
  },
});

```

**Comment by buschtoens** [comment]
Great idea!

Ideally we could solve this using [Spreadable Arguments (aka "splarguments")](https://github.com/emberjs/rfcs/pull/593), however the RFC is not even finished yet.

I think extending `{{t}}` or rather the underlying `AbstractHelper` is safe, because it currently only accepts a single `value`:

https://github.com/ember-intl/ember-intl/blob/d4c9d86aa809f9b8215e6cadad52278f8b5e7f89/addon/helpers/-format-base.js#L31-L43

We could change it to something like:

```ts
export default class AbstractHelper extends Helper {
  // ...

  compute([value, positionalOptions], namedOptions) {
    const options = Object.assign({}, positionalOptions, namedOptions);

    if (isEmpty(value)) {
      if (options.allowEmpty ?? this.allowEmpty) {
        return;
      }

      if (typeof value === 'undefined') {
        throw new Error(`${this} helper requires value attribute.`);
      }
    }

    return this.format(value, options);
  }

  // ...
}
```

Then you could invoke it like so:

```ts
{{t "my.key" (hash a="hello" b="world")}} => hello world
{{t "my.key" (hash a="hello" b="world") b="planet"}} hello planet
```

Because we're changing this in the `AbstractHelper`, this would also affect / improve the other formatting helpers, which I think is positive as well.

From my POV I would happily accept a PR, unless anyone steps forward with good objections.

**Comment by jesdavpet** [comment]
Hello @buschtoens. 👋 My team is using `ember-intl` in some Ember UI projects at @hashicorp, and we arrived at the same solution _(positional argument for options)_ that you recommended above. I'm happy to try my hand at contributing this upstream

💬 🙏 PR https://github.com/ember-intl/ember-intl/pull/1633 is my first contribution to the `ember-intl` codebase, any feedback on further changes needed to get this accepted would be welcome.

**QUESTIONS:**
  1. Since this changes the base abstract helper, are tests added to `tests/unit/helpers/t-test.ts` provide sufficient?
  2. Which branch to use a base for these changes, and which branch to PR into?
  3. Any updates to docs necessary? _(If so any pointers on where & how would be great.)_

**Comment by buschtoens** [comment]
Thanks a lot @jesdavpet! Your PR looks great. ✨

To answer your questions:

1. I believe that should be sufficient.
2. Correct, AFAIK. Didn't get to spend a lot of time in this repo in the past months unfortunately.
3. I added a comment to your PR regarding this.

**Comment by buschtoens** [comment]
🚀 This feature is now available as [`v6.0.0-beta.3`](https://github.com/ember-intl/ember-intl/releases/tag/v6.0.0-beta.3).
However, please note the discussion in #1631 regarding confusion around the current versioning and `6.x` series.

Thanks for your great contribution, @jesdavpet! ❤️

**Comment by jesdavpet** [comment]
Awesome, thanks for your help @buschtoens !!! 😃 🙌 

BTW what's your instinct about backporting this change as a minor `v5.x` release also?

**Comment by buschtoens** [comment]
I have no objections to do so. The only thing stopping me right now, is that I'm unsure, what changes are currently included in `v5.7.1` / `v5.7.2`, which the backport would have to be based on.

https://github.com/ember-intl/ember-intl/issues/1631#issuecomment-1016601882

> - Two further versions were published: `v5.7.1` & `v5.7.2`, which is currently tagged as `latest` on npm.
>   - However both are missing their tags on GitHub.
>   - They _probably_ contain the most recent fix PRs:
>     - #1609
>     - #1607
>     - #1616
>   - **They probably _also_ contain all the `v6.x` series changes.**

When I have a bit of time left today, I'll check the published package `.tar` balls and try to reconstruct from the contained code, what they're based on.

FWIW, what are the `ember-intl` and `ember-source` versions you're currently using yourself?

**Comment by jesdavpet** [comment]
> _"FWIW, what are the ember-intl and ember-source versions you're currently using yourself?"_

Apologies for my delayed response @buschtoens, missed this one because the issue's closed. 😅 

An Ember LTS upgrade is on the horizon for us, but currently we're running:
- `v5.7.0` of `ember-intl`
- `v3.24.4` of `ember-source`
